tor/forward.go

```diff
@@ -0,0 +1,299 @@
+package tor
+
+import (
+	"context"
+	"crypto"
+	"crypto/rsa"
+	"fmt"
+	"strconv"
+
+	"github.com/cretz/bine/control"
+	"github.com/cretz/bine/torutil/ed25519"
+	othered25519 "golang.org/x/crypto/ed25519"
+)
+
+// OnionForward describes a port forward to an onion service.
+type OnionForward struct {
+	// ID is the service ID for this onion service.
+	ID string
+
+	// Key is the private key for this service. It is either the set key, the
+	// generated key, or nil if asked to discard the key. If present, it is
+	// *crypto/rsa.PrivateKey (1024 bit) when Version3 is false or
+	// github.com/cretz/bine/torutil/ed25519.KeyPair when Version3 is true.
+	Key crypto.PrivateKey
+
+	// Version3 says whether or not this service is a V3 service.
+	Version3 bool
+
+	// ClientAuths is the credential set for clients. The keys are username and
+	// the values are credentials. The credentials will always be present even
+	// if Tor had to generate them.
+	ClientAuths map[string]string
+
+	// PortForwards defines the ports that will be forwarded to the onion
+	// service.
+	PortForwards PortForwardMap
+
+	// The Tor object that created this. Needed for Close.
+	Tor *Tor
+}
+
+// PortForwardMap defines local sockets to be forwarded as remote ports on the
+// onion service.
+//
+// Local sockets may be a TCP or a UNIX socket address.
+//
+// Remote ports are TCP port numbers on the onion service.
+//
+// Example:
+//   PortForwards{
+//     // Forward a local UNIX socket to port 32000 on the onion address.
+//     "/run/server.sock": []int{32000},
+//
+//     // Forward localhost port 8000 to ports 80 and 8080 on the onion
+//     // address.
+//     "127.0.0.1:8000": []int{80,8080},
+//
+//     // Forward a port from a different host on the local network.
+//     "192.168.0.1:8080": []int{8080},
+//   }
+type PortForwardMap map[string][]int
```

Comment:
Not sure we need to make this a whole separate type. Can just put `map[string][]int` and this doc directly in the `ForwardConf` struct.

Reply:
Done